packages/@glimmer/compiler/lib/template-compiler.ts

```diff
@@ -137,13 +137,14 @@ export default class TemplateCompiler {
     } else if (isNamedBlock(action)) {
       this.opcode(['closeNamedBlock', action]);
     } else if (isComponent(action)) {
-      this.opcode(['closeComponent', action], action);
-    } else if (action.modifiers.length > 0) {
       for (let i = 0; i < action.modifiers.length; i++) {
         this.modifier([action.modifiers[i]]);
       }
-      this.opcode(['closeElement', action], action);
+      this.opcode(['closeComponent', action], action);
     } else {
+      for (let i = 0; i < action.modifiers.length; i++) {
```

Comment:
@chancancode I did this, but it wasn't clear to me if the right approach was to add a modifier opcode before closing the component (like it happens when closing regular Elements) or the modifiers when applied to components should just be part of the arguments of the `openComponent`.

Reply:
@cibernox This should apply to both dynamic (`<this.foo ... />`) and regular components (`<Foo />`). Basically everything other than named blocks. So it probably makes sense to structure it like this:

```js
  closeElement([action]: [AST.ElementNode]) {
    if (isNamedBlock(action)) {
      // assert no attributes and modifiers here probably, unless that's already handled elsewhere
      this.opcode(['closeNamedBlock', action]);
    } else {
      for (let i = 0; i < action.modifiers.length; i++) { ... }

      if (isDynamicComponent(action)) {
        ...
      } else if (...) {
      } else if (...) {
      } else {
        unreachable();
      }
    }
  }
```

Reply:
as far as the factoring, the idea is that `TemplateCompiler` takes the AST (nested tree-structure), "flattens" it into an array of operations that needs to happen, and `JavaScriptCompiler` (terrible legacy name at this point) takes the falt list of operations and emit a the wire-format (mostly flat, but some "nesting" from having blocks).

So yes, this should emit the modifier opcodes before the close opcode, and the next stage can resemble them into the attributes block.